ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -72,3 +160,203 @@ func (r *RayServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		For(&rayv1alpha1.RayService{}).
 		Complete(r)
 }
+
+func (r *RayServiceReconciler) getRayServiceInstance(ctx context.Context, request ctrl.Request) (*rayv1alpha1.RayService, error) {
+	rayServiceInstance := &rayv1alpha1.RayService{}
+	if err := r.Get(ctx, request.NamespacedName, rayServiceInstance); err != nil {
+		if errors.IsNotFound(err) {
+			rayServiceLog.Info("Read request instance not found error!")
+		} else {
+			rayServiceLog.Error(err, "Read request instance error!")
+		}
+		// Error reading the object - requeue the request.
+		return nil, err
+	}
+	return rayServiceInstance, nil
+}
+
+func (r *RayServiceReconciler) updateState(ctx context.Context, rayServiceInstance *rayv1alpha1.RayService, status rayv1alpha1.ServiceStatus, err error) error {
+	rayServiceInstance.Status.ServiceStatus = status
+	if errStatus := r.Status().Update(ctx, rayServiceInstance); errStatus != nil {
+		return fmtErrors.Errorf("combined error: %v %v", err, errStatus)
+	}
+	return err
+}
+
+func (r *RayServiceReconciler) getOrCreateRayClusterInstance(ctx context.Context, rayServiceInstance *rayv1alpha1.RayService) (*rayv1alpha1.RayCluster, error) {
+	// Update ray cluster
+	rayClusterNamespacedName := types.NamespacedName{
+		Namespace: rayServiceInstance.Namespace,
+		Name:      rayServiceInstance.Name + rayClusterSuffix,
+	}
+
+	rayClusterInstance := &rayv1alpha1.RayCluster{}
+	err := r.Get(ctx, rayClusterNamespacedName, rayClusterInstance)
+
+	if err == nil {
+		rayClusterInstance.Spec = rayServiceInstance.Spec.RayClusterSpec
+
+		r.Log.Info("Update ray cluster spec")
+		if err := r.Update(ctx, rayClusterInstance); err != nil {
```

Comment:
em. I am thinking whether we should force update the cluster spec. 
If we use suffix, the conflict chance would be low. 

Another dependency is we reply on cluster upgrade to sync the latest configurations.  https://github.com/ray-project/kuberay/issues/289.

Reply:
I think the proposed v0 here is to scrap the old RayCluster and create a new one if there's a config difference.
That can be changed once the RayCluster controller supports proper in-place updates.

Reply:
I will add a TODO, in next pr will cluster creating and service switch.